stdlib/src/lib.rs

```diff
@@ -45,12 +66,125 @@ impl Default for StdLibrary {
         static STDLIB: LazyLock<StdLibrary> = LazyLock::new(|| {
             let contents =
                 Library::read_from_bytes(StdLibrary::SERIALIZED).expect("failed to read std masl!");
-            StdLibrary(contents)
+            StdLibrary::new(contents)
         });
         STDLIB.clone()
     }
 }
 
+impl HostLibrary for StdLibrary {
+    fn get_event_handlers<A>(&self) -> Vec<Box<dyn EventHandler<A>>>
+    where
+        A: AdviceProvider + 'static,
+    {
+        vec![Box::new(FalconSigToStackEventHandler::new(Box::new(
+            self.falcon_sig_event_handler.clone(),
+        )))]
+    }
+
+    fn get_mast_forest(&self) -> Arc<MastForest> {
+        self.mast_forest().clone()
+    }
+}
+
+// EVENTS
+// ================================================================================================
+
+pub const EVENT_FALCON_SIG_TO_STACK: u32 = 3419226139;
+
+// EVENT HANDLERS
+// ==============================================================================================
+
+/// An event handler which verifies a Falcon signature and pushes the result onto the stack.
+pub struct FalconSigToStackEventHandler<A> {
+    signer: Box<dyn FalconSigner<A>>,
+}
+
+impl<A> FalconSigToStackEventHandler<A> {
+    /// Creates a new instance of the Falcon signature to stack event handler, given a specified
+    /// Falcon signer.
+    pub fn new(signer: Box<dyn FalconSigner<A>>) -> Self {
+        Self { signer }
+    }
+}
+
+impl<A> Default for FalconSigToStackEventHandler<A> {
+    fn default() -> Self {
+        Self { signer: Box::new(DefaultFalconSigner) }
+    }
+}
+
+impl<A> EventHandler<A> for FalconSigToStackEventHandler<A>
+where
+    A: AdviceProvider,
+{
+    fn id(&self) -> u32 {
+        EVENT_FALCON_SIG_TO_STACK
+    }
+
+    fn on_event(
+        &mut self,
+        process: ProcessState,
+        advice_provider: &mut A,
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        let pub_key = process.get_stack_word(0);
+        let msg = process.get_stack_word(1);
+
+        let signature = self.signer.sign_message(pub_key, msg, advice_provider)?;
+
+        for r in signature {
+            advice_provider.push_stack(AdviceSource::Value(r))?;
+        }
+
+        Ok(())
+    }
+}
+
+/// A trait for signing messages using the Falcon signature scheme.
+///
+/// This trait is used by [FalconSigToStackEventHandler] to sign messages using the Falcon signature
+/// scheme.
+///
+/// It is recommended to use [dsa::falcon_sign] to implement this trait once the private key has
+/// been fetched from a user-defined location.
+pub trait FalconSigner<A>: Send + Sync {
+    /// Signs the message using the Falcon signature scheme, and returns the signature as a
+    /// `Vec<Felt>`.
+    fn sign_message(
+        &self,
+        pub_key: Word,
+        msg: Word,
+        advice_provider: &A,
+    ) -> Result<Vec<Felt>, Box<dyn Error + Send + Sync + 'static>>
+    where
+        A: AdviceProvider;
+}
```

Comment:
We are including the advice provider into the interface so that we could read the private key from it, right? I'm wondering if we could avoid this if we convert this trait into a closure - e.g., something like:
```rust
type FalconSigner = Box<dyn Fn(Word, Word) -> Vec<Felt>>;
```

And then in the `FalconSigToStackEventHandler::on_event()` we'd just invoke this closure capturing the private key we get from the advice provider.

Reply:
But this would force implementations to getting the private key from the advice provider, which is what we're trying to avoid (e.g. in `miden-base` we get it from somewhere else).

I also abstracted away the signing part to give the option to e.g. sign with a hardware device - and exposed `dsa::falcon_sign()` for convenience otherwise.

Reply:
> But this would force implementations to getting the private key from the advice provider,

No, the idea here is for the closure to take only 2 parameters: a public key and a message (the two words). The private key would be captured by the closure from the context. We are doing something similar in the [client](https://github.com/0xPolygonMiden/miden-client/blob/7d0f943c78e24469b72cda9f51d18d52b486b28c/crates/rust-client/src/sync/mod.rs#L181-L186) - though there, it is a bit more complex because the closures need to be `async`.